PR #3 for silent failure work - Add evss files and changes for uploading a document

pmclaren19 opened the pull request:
## Summary

This pr is the second pr to merge some of the work from https://github.com/department-of-veterans-affairs/vets-api/pull/20105. The first pr that included the db changes was previously merged https://github.com/department-of-veterans-affairs/vets-api/pull/20318. The second pr that added the model, factory, constants, helpers and vcr cassettes was also previously merged https://github.com/department-of-veterans-affairs/vets-api/pull/20346.

This pr does the following:

- Updated codeowners and removed the second `@department-of-veterans-affairs/benefits-management-tools-be` from  `app/sidekiq/evss/failure_notification.rb`
- Updated `app/services/evss_claim_service.rb` so that we add to the evidence_submission table when the ff `cst_send_evidence_submission_failure_emails` is enabled. Also use the helpers and constants.
-  Updated `app/sidekiq/evss/document_upload.rb` so that we updated the evidence_submission table when the ff `cst_send_evidence_submission_failure_emails` is enabled.  Also use the helpers and constants. Updated logic to pass in personalization with an obscured file name.
- Updated `app/sidekiq/evss/failure_notification.rb` to expect a personalization object instead of first_name, filename, date_submitted, and date_failed
- Updated `modules/mobile/spec/requests/mobile/v0/claim/legacy_documents_spec.rb` moved variables around, added user_account so tests passed, and updated logic so that it had the flipper `cst_send_evidence_submission_failure_emails` enabled and disabled for all tests.
- Updated `spec/requests/v0/evss_claims/documents_spec.rb`, added user_account so tests passed, and updated logic so that it had the flipper `cst_send_evidence_submission_failure_emails` enabled and disabled for all tests..
- Updated `spec/services/evss_claim_service_spec.rb` 
- Updated `spec/sidekiq/evss/document_upload_spec.rb`
- Updated `spec/sidekiq/evss/failure_notification_spec.rb`
- Updated `lib/lighthouse/benefits_documents/utilities/helpers.rb` to add `format_date_for_mailers()` 

## Related issue(s)

- Link to ticket created in va.gov-team repo https://github.com/department-of-veterans-affairs/va.gov-team/issues/100037
- Link to draft pr that this is a part of [CST]: DRAFT PR for Feature branch updated es error handling  #20105

## Testing done

- [x] *New code is covered by unit tests*
*Describe what the old behavior was prior to the change*
- The code didn't add or update records in the evidence_submissions table

Information for how to add va-notify to settings.local.yaml -> [here](teams/benefits-portfolio/benefits-management-tools/claim-status-tool/engineering/va-notify.md)

*How to test that feature flag works when ffs are disabled and we upload a file successfully* 
- Make sure `benefits_documents_use_lighthouse` is disabled
- `cst_send_evidence_failure_emails` can be disabled or enabled
- Make sure `cst_send_evidence_submission_failure_emails` is disabled
- Locally make sure your `settings.local.yml` has a vanotify section
- Update document_upload and failure_notification so that sidekiq retries are set to 0 for testing locally
- Use socks to open an ssh tunnel using port 4431
- Run vets-api and vets-website locally and upload a file
- Should see that no records were added/updated for evidence_submission table

*How to test that feature flag works when ffs are disabled and we upload a file and get an error, causing an email to be sent* 
- Make sure `benefits_documents_use_lighthouse` is disabled
- Make sure `cst_send_evidence_failure_emails` is enabled
- Make sure `cst_send_evidence_submission_failure_emails` is disabled
- Locally make sure your `settings.local.yml` has a vanotify section
- Update document_upload and failure_notification so that sidekiq retries are set to 0 for testing locally
- Update failure_notification so that in `notify_client.send_email()` we replace `recipient_identifier: { id_value: icn, id_type: 'ICN' }` with `email_address: 'YOUR_EMAIL',`
- Turn off ssh tunnel
- Run vets-api and vets-website locally and upload a file
- Should see that no records were added/updated for evidence_submission table and failure notification was called and an email was sent to you for a failed document upload

*How to test that feature flag works when ffs are enabled and we upload a file successfully* 
- Make sure `benefits_documents_use_lighthouse` is disabled
- `cst_send_evidence_failure_emails` can be disabled or enabled
- Make sure `cst_send_evidence_submission_failure_emails` is enabled
- Locally make sure your `settings.local.yml` has a vanotify section
- Update document_upload and failure_notification so that sidekiq retries are set to 0 for testing locally
- Use socks to open an ssh tunnel using port 4431
- Run vets-api and vets-website locally and upload a file
- Should see that 1 record was added/updated for evidence_submission table

*How to test that feature flag works when ffs are enabled and we upload a file and get an error, causing an email to be sent* 
- Make sure `benefits_documents_use_lighthouse` is disabled
- `cst_send_evidence_failure_emails` can be disabled or enabled
- Make sure `cst_send_evidence_submission_failure_emails` is enabled
- Locally make sure your `settings.local.yml` has a vanotify section
- Update document_upload so that sidekiq retries are set to 0 for testing locally
- Turn off ssh tunnel
- Run vets-api and vets-website locally and upload a file
- Should see that 1 record was added/updated for evidence_submission table (NOTE: in our next pr we will have an email polling job sent to you for a failed document upload and in that job you will - Update a file so that where `notify_client.send_email()` is called we replace `recipient_identifier: { id_value: icn, id_type: 'ICN' }` with `email_address: 'YOUR_EMAIL',`)

- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
**Test that feature flag works when `cst_send_evidence_failure_emails` is enabled, `benefits_documents_use_lighthouse` is disabled and `cst_send_evidence_submission_failure_emails` is disabled. When you upload a file, no evidence submission record is added and file is successfully uploaded:**
![Screenshot 2025-01-22 at 9 24 47 AM](https://github.com/user-attachments/assets/0c281d0e-1014-4763-8d42-91097327114a)
![Screenshot 2025-01-22 at 9 42 13 AM](https://github.com/user-attachments/assets/1c020b12-c5cb-4c03-bfcc-e37b04f9b427)
![Screenshot 2025-01-22 at 10 08 45 AM](https://github.com/user-attachments/assets/c83ce209-b6e8-4bef-b325-b128234c037c)
![Screenshot 2025-01-22 at 9 27 26 AM](https://github.com/user-attachments/assets/818f572d-e886-4fde-ae58-c0e373e72992)

**Test that feature flag works when `cst_send_evidence_failure_emails` is enabled, `benefits_documents_use_lighthouse` is disabled and `cst_send_evidence_submission_failure_emails` is disabled. When you upload a file and an error occurs, no evidence submission record is added and an email is sent for failed upload**
![Screenshot 2025-01-22 at 1 02 01 PM](https://github.com/user-attachments/assets/b1a7f926-5609-44d2-a032-54f9db0f1197)
![Screenshot 2025-01-22 at 12 59 33 PM](https://github.com/user-attachments/assets/b6217bfb-fd85-4009-b21a-488a6c2b59cf)

**Test that feature flag works when `cst_send_evidence_failure_emails` is enabled or disabled, `benefits_documents_use_lighthouse` is disabled and `cst_send_evidence_submission_failure_emails` is enabled. When you upload a file, an evidence submission record is added and file is successfully uploaded:**
![Screenshot 2025-01-22 at 11 42 32 AM](https://github.com/user-attachments/assets/a18b0c61-5354-40cc-a8bc-4a5efb078d8b)
![Screenshot 2025-01-22 at 11 44 12 AM](https://github.com/user-attachments/assets/c3e9b91e-d971-4df5-a766-e63b8f174823)
![Screenshot 2025-01-22 at 11 42 50 AM](https://github.com/user-attachments/assets/a2d4a176-7410-4094-b1dd-57bf323374ed)

**Test that feature flag works when `cst_send_evidence_failure_emails` is enabled or disabled, `benefits_documents_use_lighthouse` is disabled and `cst_send_evidence_submission_failure_emails` is enabled. When you upload a file, an evidence submission record is added and updated (it will have a FAILED upload_status and a failed_date) and no email is sent for failed upload (NOTE: an email will eventually be sent when the polling job runs):**
![Screenshot 2025-01-23 at 8 48 51 AM](https://github.com/user-attachments/assets/571571dc-6e4f-4165-8c9a-e669f2b153d0)
![Screenshot 2025-01-23 at 8 50 09 AM](https://github.com/user-attachments/assets/92dd34fc-3fc5-48a6-bdfb-6282cd36b14d)


## What areas of the site does it impact?
Claim Status Tool

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

RachalCassity commented:
@pmclaren19 Curious, why only add the vanotify section to settings.local.yml? Is there documentation on what is needed in settings.local.yml?

jerekshoe commented:
Also, I believe Peri has the needed info documented, but it uses our teams staging credentials which we're not supposed to share out publicly

RachalCassity commented on spec/services/evss_claim_service_spec.rb:
```diff
@@ -132,6 +140,24 @@
+    end
+
+    context 'when :cst_send_evidence_submission_failure_emails is disabled' do
+      before { Flipper.disable(:cst_send_evidence_submission_failure_emails) }
```
Please change all Flipper.enable and Flipper.disable to this format to prevent flakey tests:

 allow(Flipper).to receive(:enabled?).with(:cst_send_evidence_submission_failure_emails).and_return(true) OR
allow(Flipper).to receive(:enabled?).with(: cst_send_evidence_submission_failure_emails).and_return(false)

pmclaren19 replied:
@RachalCassity this causes tests to fail when I try to add what you put here...

pmclaren19 replied:
it was a formatting issue with how the flipper was declared in the code. All resolved. I updated `spec/services/evss_claim_service_spec.rb` and `spec/sidekiq/evss/document_upload_spec.rb` per your request.  Please re-review when you have time.

pmclaren19 replied:
After some digging I found that when we use the new flipper logic, if there is another flipper in the code that is not set within the test, then the test will fail. We have to set all flippers that the test touches in a before loop. Issue should now be fixed!

RachalCassity commented on modules/mobile/spec/requests/mobile/v0/claim/legacy_documents_spec.rb:
```diff
@@ -5,21 +5,24 @@
 
+  before do
+    Flipper.disable(:mobile_lighthouse_document_upload)
+    Flipper.disable(:mobile_lighthouse_document_upload, user)
```
Please remove Line 20 extra line.

pmclaren19 commented:
@jerekshoe @samcoforma @RachalCassity please review again! I updated some tests, added logging to document_upload when a files is successfully uploaded, created a new helper function `format_date_for_mailers` and fixed a bug with the failed date that we were storing in the personalization field. Please let me know if you have any questions.

RachalCassity commented:
Approving this pull request and merging because Danger.
